File: typescript/packages/sdk/src/command.ts
import { hash } from "./canonical.ts";
import { buildManifest } from "./manifest.ts";
import type {
  Manifest,
  JsonSchema,
  Effects,
  Determinism,
  Sandbox,
  CommandBlock,
  ArgvTemplate,
} from "./types.ts";

/**
 * Command / CLI capabilities (SPEC §28). A `command` capability is a
 * content-addressed, argv-typed CLI invocation that is NEVER run through a
 * shell. The argv is built from an `argv_template` of literal tokens and typed
 * `{param, schema}` holes; each parameter value occupies exactly ONE argv
 * element and is never re-split, re-quoted, globbed, or shell-expanded. This
 * eliminates CWE-78 shell injection by construction (§28.1).
 */

/**
 * Resolve an argv_template against params into a flat string[] (SPEC §28.1).
 *
 * - A literal string token passes through verbatim.
 * - A {param, schema} hole is replaced by EXACTLY ONE argv element: the param's
 *   value, stringified, never split or quoted. A value like "; rm -rf / #"
 *   becomes a single literal argv element — never a new command.
 *
 * No shell is ever consulted; there is no interpolation, globbing, or
 * word-splitting. Missing params throw rather than silently producing a hole.
 */
export function resolveArgv(
  argvTemplate: ArgvTemplate,
  params: Record<string, unknown>,
): string[] {
  const argv: string[] = [];
  for (const token of argvTemplate) {
    if (typeof token === "string") {
      // Literal token: passes through, one argv element.
      argv.push(token);
      continue;
    }
    // Typed hole: exactly one argv element, never re-split or expanded.
    const { param } = token;
    if (!Object.prototype.hasOwnProperty.call(params, param)) {
      throw new Error(`resolveArgv: missing required param "${param}"`);
    }
    const value = params[param];
    if (value === null || value === undefined) {
      throw new Error(`resolveArgv: param "${param}" is null/undefined`);
    }
    // One value → one argv slot. String/number/boolean stringify in place; an
    // object/array is rejected (it would be ambiguous to flatten and could be a
    // smuggling attempt).
    if (typeof value === "object") {
      throw new Error(
        `resolveArgv: param "${param}" must be a scalar; objects/arrays cannot occupy a single argv slot`,
      );
    }
    argv.push(String(value));
  }
  return argv;
}

/**
 * argv_hash = sha256(JCS(resolved_argv)) (SPEC §7, §28.1). The grant binds this
 * over the fully-resolved argv ARRAY; a hijacked Planner cannot add, remove, or
 * alter a token after approval without invalidating the grant
 * (ARGUMENT_HASH_MISMATCH).
 */
export function argvHash(argv: string[]): string {
  return hash(argv);
}

export interface BuildCommandManifestInput {
  issuer: string;
  provider: string;
  name: string;
  version: string;
  summary_for_user: string;
  summary_for_model: string;
  /** input_schema for the typed parameters (additionalProperties:false, §28.1.4). */
  input_schema: JsonSchema;
  output_schema: JsonSchema;
  effects: Effects;
  determinism: Determinism;
  sandbox: Sandbox;
  /** The §28 command block. `shell` is forced to false. */
  command: Omit<CommandBlock, "shell"> & { shell?: false };
}

/**
 * Build an UNSIGNED `command` capability manifest with a correct
 * content-addressed identity (SPEC §28, §4.1).
 *
 * CRITICAL (§4.1): for a command capability the `command` block is part of the
 * contract — it is APPENDED to the eight common contract fields before hashing.
 * Two capabilities identical but for `command.exec_digest` (or any other
 * command field) therefore get different `contract_hash` / `capability_id`.
 * This is the rug-pull defense for bridged binaries (§28.4).
 */
export function buildCommandManifest(
  input: BuildCommandManifestInput,
): Omit<Manifest, "signature"> {
  const command: CommandBlock = {
    binary: input.command.binary,
    shell: false,
    argv_template: input.command.argv_template,
    ...(input.command.exec_digest !== undefined
      ? { exec_digest: input.command.exec_digest }
      : {}),
    ...(input.command.working_dir !== undefined
      ? { working_dir: input.command.working_dir }
      : {}),
    ...(input.command.provenance !== undefined
      ? { provenance: input.command.provenance }
      : {}),
    ...(input.command.subcommand_allow !== undefined
      ? { subcommand_allow: input.command.subcommand_allow }
      : {}),
  };

  // buildManifest computes identity over the eight common fields. For a command
  // capability we must additionally fold the command block into the contract.
  // We pass the command block through `command` so buildManifest can append it.
  return buildManifest({
    issuer: input.issuer,
    provider: input.provider,
    name: input.name,
    version: input.version,
    summary_for_user: input.summary_for_user,
    summary_for_model: input.summary_for_model,
    input_schema: input.input_schema,
    output_schema: input.output_schema,
    effects: input.effects,
    determinism: input.determinism,
    sandbox: input.sandbox,
    kind: "command",
    command,
  });
}

export interface BridgeExistingCliInput {
  binary: string;
  /** Pinned hash of the resolved executable. A changed binary is a new identity (§28.4). */
  execDigest: string;
  /** The allowed subcommand/flag patterns, as a signed contract (not host-local settings). */
  subcommandAllow: string[];
  argvTemplate: ArgvTemplate;
  /** Identity / signing fields. Sensible defaults are supplied for a bridge. */
  issuer?: string;
  provider?: string;
  name?: string;
  version?: string;
  effects?: Effects;
  determinism?: Determinism;
  sandbox?: Sandbox;
  input_schema?: JsonSchema;
  output_schema?: JsonSchema;
  working_dir?: string;
}

/**
 * The command bridge (SPEC §28.4): turn an ordinary existing CLI that has no
 * VCP manifest into a constrained `command` capability WITHOUT modifying it.
 *
 * A bridge MUST: pin the binary's identity by `exec_digest` (rug-pull defense,
 * §4 / §28.4); express the allowlist as a signed contract (`subcommand_allow`),
 * not host-local settings; apply argv-only execution; and mark provenance
 * `host_cli`. The returned manifest carries `provenance:"host_cli"` and the
 * pinned exec_digest INSIDE the (identity-bearing) command block, so a changed
 * binary digest yields a new, unapproved capability.
 */
export function bridgeExistingCli(
  input: BridgeExistingCliInput,
): Omit<Manifest, "signature"> {
  const name = input.name ?? `host_cli.${input.binary}`;
  // Derive parameter names from the typed holes so the input_schema is strict
  // (additionalProperties:false, §28.1.4) by default.
  const holeParams = input.argvTemplate
    .filter((t): t is { param: string; schema: JsonSchema } => typeof t !== "string")
    .map((t) => t.param);
  const properties: Record<string, unknown> = {};
  for (const t of input.argvTemplate) {
    if (typeof t !== "string") properties[t.param] = t.schema;
  }
  const defaultInputSchema: JsonSchema = {
    type: "object",
    additionalProperties: false,
    properties,
    required: holeParams,
  };

  // Conservative defaults: a bridged write requires approval; a read-only host
  // CLI can be narrowed by an operator. Default to write-irreversible+approval.
  const defaultEffects: Effects = {
    class: "write-irreversible",
    external_side_effect: true,
    requires_user_approval: true,
  };

  return buildCommandManifest({
    issuer: input.issuer ?? "did:web:host.local",
    provider: input.provider ?? "host.cli",
    name,
    version: input.version ?? "0.0.0-host_cli",
    summary_for_user: `Bridged host CLI "${input.binary}" exposed as a constrained, digest-pinned VCP command (§28.4).`,
    summary_for_model:
      `Bridged command "${input.binary}". argv-only, no shell. ` +
      `Allowed subcommands/flags: ${input.subcommandAllow.join(", ") || "none"}. ` +
      `Parameters: ${holeParams.join(", ") || "none"}.`,
    input_schema: input.input_schema ?? defaultInputSchema,
    output_schema: input.output_schema ?? { type: "object" },
    effects: input.effects ?? defaultEffects,
    determinism: input.determinism ?? { class: "nondeterministic" },
    sandbox: input.sandbox ?? { filesystem: "none", network: [], secrets: [] },
    command: {
      binary: input.binary,
      exec_digest: input.execDigest,
      argv_template: input.argvTemplate,
      provenance: "host_cli",
      subcommand_allow: input.subcommandAllow,
      ...(input.working_dir !== undefined ? { working_dir: input.working_dir } : {}),
    },
  });
}

File: go/sdk/command.go
package sdk

import (
	"encoding/json"
	"fmt"
)

// command.go implements the §28 command/CLI capability: the argv model (no shell,
// ever), the identity-bearing `command` block appended to the contract (§4.1,
// §28.4), and the command bridge for wrapping an existing host CLI (§28.4).
//
// The single most important property here is structural: a `command` capability is
// executed by directly exec'ing `binary` with an argv ARRAY built from the
// template. There is no shell, no interpolation, no globbing, no word-splitting. A
// parameter value such as "; rm -rf / #" becomes ONE literal argv element passed to
// the program, never a new command — CWE-78 shell injection is eliminated by
// construction (§28.1). The real executor lives in the gateway package
// (gateway.BuildCommandExec / RunCommand) so that os/exec stays out of the SDK; the
// resolution and hashing primitives that bind the argv to a grant live here.

// ArgvToken is one element of a command's argv_template (spec §28.1). It is either
// a literal token (a fixed string copied verbatim into argv) or a typed hole that
// names a parameter and carries the JSON Schema constraining that parameter's
// value. Exactly one of Literal / Param is meaningful per token: IsParam reports
// which.
//
// Modeled as a struct rather than an interface so it JSON-marshals cleanly and so
// the literal-vs-hole decision is a single boolean rather than a type switch. A
// custom (Un)marshaler maps it to the schema's oneOf: a literal is a bare JSON
// string; a hole is a {"param","schema"} object.
type ArgvToken struct {
	// Literal is the fixed argv element when this token is not a parameter hole.
	Literal string
	// Param is the parameter name when this token IS a typed hole.
	Param string
	// Schema is the JSON Schema for the parameter value (typed hole only). Its
	// presence does not affect argv resolution — resolution substitutes the value as
	// a single element regardless — but it carries the vcp_kind ("path"/"host") and
	// type constraints the Gateway enforces before resolving (§28.1 rule 4).
	Schema any
	// isParam records whether this token is a typed hole. It is set by the
	// constructors and by UnmarshalJSON; callers use IsParam to read it.
	isParam bool
}

// LiteralToken builds a literal argv token (a fixed string element).
func LiteralToken(s string) ArgvToken {
	return ArgvToken{Literal: s, isParam: false}
}

// ParamToken builds a typed-hole argv token naming a parameter and its schema.
func ParamToken(param string, schema any) ArgvToken {
	return ArgvToken{Param: param, Schema: schema, isParam: true}
}

// IsParam reports whether this token is a typed parameter hole (true) or a literal
// argv element (false).
func (t ArgvToken) IsParam() bool { return t.isParam }

// MarshalJSON renders a literal token as a bare JSON string and a typed hole as a
// {"param","schema"} object, matching the manifest schema's argv_template oneOf.
func (t ArgvToken) MarshalJSON() ([]byte, error) {
	if t.isParam {
		obj := map[string]any{"param": t.Param}
		if t.Schema != nil {
			obj["schema"] = t.Schema
		}
		return json.Marshal(obj)
	}
	return json.Marshal(t.Literal)
}

// UnmarshalJSON accepts either a bare string (literal token) or a {param,schema}
// object (typed hole), per the manifest schema's argv_template oneOf.
func (t *ArgvToken) UnmarshalJSON(b []byte) error {
	// Try a bare string first (literal token).
	var s string
	if err := json.Unmarshal(b, &s); err == nil {
		t.Literal = s
		t.Param = ""
		t.Schema = nil
		t.isParam = false
		return nil
	}
	var obj struct {
		Param  string `json:"param"`
		Schema any    `json:"schema"`
	}
	if err := json.Unmarshal(b, &obj); err != nil {
		return fmt.Errorf("command: argv token is neither a string nor a {param,schema} object: %w", err)
	}
	if obj.Param == "" {
		return fmt.Errorf("command: argv token object missing non-empty \"param\"")
	}
	t.Literal = ""
	t.Param = obj.Param
	t.Schema = obj.Schema
	t.isParam = true
	return nil
}

// Command is a manifest's content-addressed, argv-typed `command` block (spec §28,
// schemas/manifest.schema.json). It is NEVER executed via a shell; the Gateway
// exec's Binary with the resolved argv array.
//
// Because the command block determines what actually runs, it is identity-bearing:
// it is appended to the contract before hashing (§4.1, §28.4), so a changed
// ExecDigest or ArgvTemplate yields a different capability_id.
type Command struct {
	// Binary is the executable path or name (spec §28.1).
	Binary string `json:"binary"`
	// ExecDigest is the pinned sha256 of the resolved executable (§28.4). A changed
	// binary on disk no longer matches and is a new, unapproved identity. Optional in
	// the schema; omitted when empty.
	ExecDigest string `json:"exec_digest,omitempty"`
	// Shell MUST be false. VCP never passes commands to a shell (§28.1). It is a
	// typed field rather than implicit so the false value is part of the signed
	// contract and cannot silently flip to true.
	Shell bool `json:"shell"`
	// ArgvTemplate is the ordered list of literal tokens and typed holes (§28.1).
	ArgvTemplate []ArgvToken `json:"argv_template"`
	// WorkingDir is the working directory; MUST be within sandbox.filesystem (§28.2).
	WorkingDir string `json:"working_dir,omitempty"`
	// Provenance is "authored" (default) or "host_cli" (a bridged existing CLI, §28.4).
	Provenance string `json:"provenance,omitempty"`
	// SubcommandAllow is, for bridged CLIs, the allowed subcommand/flag patterns as a
	// signed contract rather than host-local settings (§28.4).
	SubcommandAllow []string `json:"subcommand_allow,omitempty"`
}

// Provenance values for a command capability (spec §28.4).
const (
	ProvenanceAuthored = "authored"
	ProvenanceHostCLI  = "host_cli"
)

// asMap renders the command block as a map[string]any for canonicalization,
// omitting empty optional fields exactly as the JSON tags do, so the hashed form
// matches the wire form. shell is always present (and always false). Built
// explicitly rather than via reflection so identity is exactly the declared fields.
func (c Command) asMap() (map[string]any, error) {
	m := map[string]any{
		"binary": c.Binary,
		"shell":  c.Shell,
	}
	if c.ExecDigest != "" {
		m["exec_digest"] = c.ExecDigest
	}
	if c.WorkingDir != "" {
		m["working_dir"] = c.WorkingDir
	}
	if c.Provenance != "" {
		m["provenance"] = c.Provenance
	}
	if len(c.SubcommandAllow) > 0 {
		sa := make([]any, len(c.SubcommandAllow))
		for i, s := range c.SubcommandAllow {
			sa[i] = s
		}
		m["subcommand_allow"] = sa
	}
	// argv_template is always present (it defines what runs). Each token marshals to
	// a string (literal) or a {param,schema} object (hole). Round-trip through JSON
	// so nested schema values become the float64/map shapes Canonicalize expects.
	toks := make([]any, len(c.ArgvTemplate))
	for i, tok := range c.ArgvTemplate {
		v, err := tok.toCanonicalValue()
		if err != nil {
			return nil, err
		}
		toks[i] = v
	}
	m["argv_template"] = toks
	return m, nil
}

// toCanonicalValue renders one token as the decoded JSON value (string or map) that
// Canonicalize accepts, so the token participates in the contract hash identically
// to its wire form.
func (t ArgvToken) toCanonicalValue() (any, error) {
	raw, err := t.MarshalJSON()
	if err != nil {
		return nil, err
	}
	var v any
	if err := json.Unmarshal(raw, &v); err != nil {
		return nil, err
	}
	return v, nil
}

// ResolveArgv resolves an argv_template against a parameter map into the concrete
// argv array passed to exec (spec §28.1). Each literal token becomes exactly one
// element copied verbatim; each typed hole becomes exactly ONE element whose value
// is the parameter's string value — never split, re-quoted, globbed, or
// shell-expanded (§28.1 rules 1, 2). A value such as "; rm -rf / #" therefore
// occupies a single argv slot and is passed to the program as one literal string.
//
// params values are taken as their string form. Only string-valued parameters are
// supported here (the command surface of VCP v0.1); a missing parameter is an error
// so resolution fails closed rather than silently dropping or empty-substituting a
// slot.
func ResolveArgv(template []ArgvToken, params map[string]string) ([]string, error) {
	argv := make([]string, 0, len(template))
	for i, tok := range template {
		if tok.IsParam() {
			val, ok := params[tok.Param]
			if !ok {
				return nil, fmt.Errorf("command: argv_template hole %d references missing parameter %q", i, tok.Param)
			}
			// Exactly one element. No splitting, no expansion (§28.1 rule 2).
			argv = append(argv, val)
			continue
		}
		argv = append(argv, tok.Literal)
	}
	return argv, nil
}

// ArgvHash computes the argument_hash a grant binds for a command capability: the
// JCS hash over the fully-resolved argv ARRAY (spec §28.1 rule 3). Binding the
// resolved argv means a hijacked Planner cannot add, remove, or alter a token after
// approval without invalidating the grant (ARGUMENT_HASH_MISMATCH). The argv is an
// array of strings, so this is HashJCS over a []any of those strings — identical to
// the language-agnostic command.json `argument_hash`.
func ArgvHash(argv []string) (string, error) {
	arr := make([]any, len(argv))
	for i, s := range argv {
		arr[i] = s
	}
	return HashJCS(arr)
}

// CommandContract is the identity-bearing contract for a command capability (spec
// §4.1, §28.4): the eight common contract fields PLUS the `command` block appended
// as a ninth member. Because JCS sorts keys, only the member SET and values matter;
// appending the command block means two capabilities identical but for their
// ExecDigest (or any argv token) compute DIFFERENT contract_hash values, so a
// changed binary is a new, unapproved capability (INV-2, security test #22).
//
// The base contract's eight fields are taken from c (the common Contract); cmd is
// the command block. The result is the map Canonicalize hashes.
func CommandContract(c Contract, cmd Command) (map[string]any, error) {
	m := c.asMap() // exactly the eight common fields
	cm, err := cmd.asMap()
	if err != nil {
		return nil, err
	}
	m["command"] = cm
	return m, nil
}

// CommandContractHash computes contract_hash = sha256(JCS(contract)) for a command
// capability, where the contract is the eight common fields + the command block
// (spec §4.1, §28.4).
func CommandContractHash(c Contract, cmd Command) (string, error) {
	mp, err := CommandContract(c, cmd)
	if err != nil {
		return "", err
	}
	return HashJCS(mp)
}

// CommandCapabilityID computes capability_id = "vcp:cap:" + name + "@" +
// contract_hash for a command capability (spec §4, §4.1).
func CommandCapabilityID(c Contract, cmd Command) (string, error) {
	h, err := CommandContractHash(c, cmd)
	if err != nil {
		return "", err
	}
	return fmt.Sprintf("vcp:cap:%s@%s", c.Name, h), nil
}

// BridgeExistingCLI wraps an existing host binary as a constrained `command`
// capability without modifying it (the command bridge, spec §28.4). It is the
// command-line analogue of BridgeMCPTool (§16): the binary's identity is pinned by
// execDigest, the allowlist is expressed as a signed contract (argvTemplate +
// subcommandAllow) rather than host-local settings, provenance is marked host_cli,
// and §28.1–28.3 apply in full (argv-only execution, sandbox, effect class).
//
// The returned manifest is UNSIGNED (a bridge holds no upstream key); the bridge
// Gateway signs it with its own key, exactly as for the MCP bridge. effectClass
// lets the caller declare the command's effect (read-only commands MAY auto-run;
// writes are gated by policy + plan/apply). sandboxFilesystem is the filesystem
// allowlist (paths) the command runs under (§28.2); networkAllow is the egress
// allowlist (empty ⇒ deny all).
func BridgeExistingCLI(
	issuer, provider, name, version string,
	binary, execDigest string,
	argvTemplate []ArgvToken,
	subcommandAllow []string,
	workingDir string,
	effectClass string,
	inputSchema, outputSchema any,
	sandboxFilesystem []string,
	networkAllow []string,
) (Manifest, error) {
	if name == "" {
		return Manifest{}, fmt.Errorf("bridge: command capability has empty name")
	}
	if binary == "" {
		return Manifest{}, fmt.Errorf("bridge: command capability has empty binary")
	}
	if execDigest == "" {
		return Manifest{}, fmt.Errorf("bridge: host_cli bridge MUST pin exec_digest (§28.4)")
	}
	if _, err := DigestHex(execDigest); err != nil {
		return Manifest{}, fmt.Errorf("bridge: exec_digest is not a sha256 digest: %w", err)
	}
	if effectClass == "" {
		effectClass = "read-only"
	}
	if version == "" {
		version = "host_cli"
	}

	cmd := Command{
		Binary:          binary,
		ExecDigest:      execDigest,
		Shell:           false,
		ArgvTemplate:    argvTemplate,
		WorkingDir:      workingDir,
		Provenance:      ProvenanceHostCLI,
		SubcommandAllow: subcommandAllow,
	}

	if inputSchema == nil {
		inputSchema = map[string]any{"type": "object", "additionalProperties": false}
	}
	if outputSchema == nil {
		outputSchema = map[string]any{"type": "object"}
	}

	fsAllow := make([]any, len(sandboxFilesystem))
	for i, p := range sandboxFilesystem {
		fsAllow[i] = p
	}
	net := make([]any, len(networkAllow))
	for i, n := range networkAllow {
		net[i] = n
	}
	var fsValue any
	if len(sandboxFilesystem) == 0 {
		fsValue = "none"
	} else {
		fsValue = fsAllow
	}
	sandbox := map[string]any{
		"filesystem": fsValue,
		"network":    net,
		"secrets":    []any{},
	}

	effects := map[string]any{
		"class":                effectClass,
		"external_side_effect": effectClass != "read-only" && effectClass != "propose-only",
	}
	determinism := map[string]any{
		"class": commandDeterminismFor(effectClass),
	}

	cmdMap, err := cmd.asMap()
	if err != nil {
		return Manifest{}, err
	}

	capability := Capability{
		Name:            name,
		Version:         version,
		SummaryForUser:  fmt.Sprintf("Bridged host CLI %q (binary %s).", name, binary),
		SummaryForModel: fmt.Sprintf("Bridged command capability %q (effect: %s). argv is typed and executed without a shell; call only within the approved plan and declared schema.", name, effectClass),
		InputSchema:     inputSchema,
		OutputSchema:    outputSchema,
		Effects:         effects,
		Determinism:     determinism,
		Sandbox:         sandbox,
		Kind:            "command",
		Command:         cmdMap,
	}

	m := NewManifest(issuer, provider, capability)
	// Identity includes the command block (§4.1); ComputeIdentity uses the
	// command-aware path because Capability.Command is populated.
	if _, _, err := m.ComputeIdentity(); err != nil {
		return Manifest{}, err
	}
	return m, nil
}

// commandDeterminismFor picks a reasonable determinism class for a bridged command
// by its effect class (spec §28.3): reads are external-read; writes that an
// idempotency key makes safe are idempotent-write. Callers who know better may
// override before signing.
func commandDeterminismFor(effectClass string) string {
	switch effectClass {
	case "write-idempotent", "write-reversible", "write-irreversible":
		return "idempotent-write"
	default:
		return "external-read"
	}
}
